Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp

```diff
@@ -0,0 +1,349 @@
+            codec = kAudioFormatMPEG4AAC_HE;
+        else if (codecName.startsWith("mp4a.40.23"_s))
+            codec = kAudioFormatMPEG4AAC_LD;
+        else if (codecName.startsWith("mp4a.40.29"_s))
+            codec = kAudioFormatMPEG4AAC_HE_V2;
+        else if (codecName.startsWith("mp4a.40.39"_s))
+            codec = kAudioFormatMPEG4AAC_ELD;
+        else
+            codec = kAudioFormatMPEG4AAC;
+    } else if (codecName == "mp3"_s)
+        codec = kAudioFormatMPEGLayer3;
+    else if (codecName == "opus"_s) {
+        codec = kAudioFormatOpus;
+    } else if (codecName == "alaw"_s)
+        codec = kAudioFormatALaw;
+    else if (codecName == "ulaw"_s)
+        codec = kAudioFormatULaw;
+    else if (codecName == "flac"_s) {
+        codec = kAudioFormatFLAC;
+    } else if (codecName == "vorbis"_s) {
+        codec = 'vorb';
+    } else if (codecName.startsWith("pcm-"_s)) {
+        auto components = codecName.toString().split('-');
+        if (components.size() != 2)
+            return makeUnexpected(makeString("Invalid LPCM codec string:"_s, codecName));
+        auto pcmFormat = components[1].convertToASCIILowercase();
+        codec = kAudioFormatLinearPCM;
+        if (pcmFormat == "u8"_s)
+            format = AudioStreamDescription::Uint8;
+        else if (pcmFormat == "s16"_s)
+            format = AudioStreamDescription::Int16;
+        else if (pcmFormat == "s24"_s)
+            format = AudioStreamDescription::Int24;
+        else if (pcmFormat == "s32"_s)
+            format = AudioStreamDescription::Int32;
+        else if (pcmFormat == "f32"_s)
+            format = AudioStreamDescription::Float32;
+        else
+            return makeUnexpected(makeString("Invalid LPCM codec format:"_s, pcmFormat));
+    } else
+        return makeUnexpected("not supported codec"_s);
+
+    return std::make_pair(codec, format);
+}
+
+Ref<AudioDecoder::DecodePromise> AudioDecoderCocoa::decode(EncodedData&& data)
+{
+    return invokeAsync(queueSingleton(), [data = SharedBuffer::create(data.data), isKeyFrame = data.isKeyFrame, timestamp = data.timestamp, duration = data.duration, decoder = m_internalDecoder]() mutable {
+        return decoder->decode(WTFMove(data), isKeyFrame, timestamp, duration);
+    });
+}
+
+Ref<GenericPromise> AudioDecoderCocoa::flush()
+{
+    return invokeAsync(queueSingleton(), [decoder = m_internalDecoder] {
+        return decoder->flush();
+    });
+}
+
+void AudioDecoderCocoa::reset()
+{
+    // FIXME: that can't be correct
+    m_internalDecoder->close();
+}
+
+void AudioDecoderCocoa::close()
+{
+    queueSingleton().dispatch([decoder = m_internalDecoder] {
+        decoder->close();
+    });
+}
+
+void InternalAudioDecoderCocoa::decompressedAudioOutputBufferCallback(void* object, CMBufferQueueTriggerToken)
+{
+    // We can only be called from the CoreMedia callback if we are still alive.
+    RefPtr decoder = static_cast<class InternalAudioDecoderCocoa*>(object);
```

Comment:
This is not totally clear how this is guaranteed here.
Can you clarify this?

For instance, `addSampleBuffer` is asynchronous and can run for some time.
What happens if the `AudioDecoderCocoa` is destroyed without close being called?

Maybe there should be a ref in addSampleBuffer lambda, or there is a need to do something in `AudioDecoderCocoa::~AudioDecoderCocoa()`?

Reply:
I do not think there is such a guarantee, say in case of stopActiveDOMObjects.
In any case, I think `AudioDecoder/InternalAudioDecoderCocoa` should be self protective, otherwise we might change slightly things in WebCodecsAudioDecoder and break the assumption that `InternalAudioDecoderCocoa` is making.

Reply:
I've made it more explicit. Calling close() in the destructor which holds a reference to itself until we're guaranteed that the converter won't output new packets

Reply:
Maybe add a comment in the destructor?